model.py

```diff
@@ -9021,13 +9047,71 @@ class Collection(Base):
 
     def __repr__(self):
         return (u'<Collection "%s"/"%s" ID=%d>' %
-                (self.name, self.protocol, self.id)).encode('utf8')
+                (self.name, self.protocol, self.id)).encode('utf8')        
+
+    @classmethod
+    def by_name_and_protocol(cls, _db, name, protocol):
+        """Find or create a Collection with the given name and the given
+        protocol.
+
+        We can't use get_one_or_create because the protocol is kept in
+        a separate database object, (an ExternalIntegration).
+
+        :return: A 2-tuple (collection, is_new)
+        """
+        qu = cls.by_protocol(_db, protocol)
+        qu = qu.filter(Collection.name==name)
+        try:
+            collection = qu.one()                    
+            is_new = False
+        except NoResultFound, e:
+            # Make a new Collection.
+            collection, is_new = get_one_or_create(_db, Collection, name=name)
+            if not is_new and collection.protocol != protocol:
+                # The collection already exists, it just uses a different
+                # protocol than the one we asked about.
+                raise ValueError(
+                    'Collection "%s" does not use protocol "%s".' % (
+                        name, protocol
+                    )
+                )
+            collection.external_integration.type=ExternalIntegration.LICENSE_TYPE
+            collection.external_integration.protocol=protocol           
+        return collection, is_new
+    
+    @classmethod
+    def by_protocol(cls, _db, protocol):
+        """Query collections that get their licenses through the given protocol.
 
+        :param protocol: Protocol to use. If this is None, all
+        Collections will be returned.
+        """
+        qu = _db.query(Collection)
+        if protocol:
+            qu = qu.join(
+            ExternalIntegration,
+            ExternalIntegration.id==Collection.external_integration_id).filter(
+                ExternalIntegration.type==ExternalIntegration.LICENSE_TYPE
+            ).filter(ExternalIntegration.protocol==protocol)
+        return qu
+    
+    @property
+    def protocol(self):
+        """What protocol do we need to use to get licenses for this 
+        collection?
+        """
+        if self.external_integration.protocol:
+            return self.external_integration.protocol
+        if self.parent:
+            return self.parent.protocol
```

Comment:
Inheriting protocol from the parent is inconsistent with by_protocol and by_name_and_protocol. A Collection with a protocol from its parent won't be able to be looked up by that protocol.